client/relay-chain-rpc-interface/src/reconnecting_ws_client.rs

```diff
@@ -0,0 +1,495 @@
+	/// The worker does the following:
+	/// - Listen for [`RpcDispatcherMessage`], perform requests and register new listeners for the notification streams
+	/// - Distribute incoming import, best head and finalization notifications to registered listeners.
+	///   If an error occurs during sending, the receiver has been closed and we remove the sender from the list.
+	/// - Find a new valid RPC server to connect to in case the websocket connection is terminated.
+	///   If the worker is not able to connec to an RPC server from the list, the worker shuts down.
+	async fn run(mut self) {
+		let mut pending_requests = FuturesUnordered::new();
+
+		let urls = std::mem::take(&mut self.ws_urls);
+		let Ok(mut client_manager) = ClientManager::new(urls).await else {
+			tracing::error!(target: LOG_TARGET, "No valid RPC url found. Stopping RPC worker.");
+			return;
+		};
+		let Ok(mut subscriptions) = client_manager.get_subscriptions().await else {
+			tracing::error!(target: LOG_TARGET, "Not able to fetch subscription after reconnect.");
+			return;
+		};
+
+		let mut imported_blocks_cache =
+			LruCache::new(NonZeroUsize::new(40).expect("40 is nonzero; qed."));
+		let mut should_reconnect = false;
+		let mut last_seen_finalized_num: RelayBlockNumber = 0;
+		loop {
+			// This branch is taken if the websocket connection to the current RPC server is closed.
+			if should_reconnect {
+				match self.handle_reconnect(&mut client_manager, &mut pending_requests).await {
+					Ok(new_subscriptions) => {
+						subscriptions = new_subscriptions;
+					},
+					Err(message) => {
+						tracing::error!(
+							target: LOG_TARGET,
+							message,
+							"Unable to reconnect, stopping worker."
+						);
+						return
+					},
+				}
+				should_reconnect = false;
+			};
+
+			tokio::select! {
+				evt = self.client_receiver.recv() => match evt {
+					Some(RpcDispatcherMessage::RegisterBestHeadListener(tx)) => {
+						self.best_header_listeners.push(tx);
+					},
+					Some(RpcDispatcherMessage::RegisterImportListener(tx)) => {
+						self.imported_header_listeners.push(tx)
+					},
+					Some(RpcDispatcherMessage::RegisterFinalizationListener(tx)) => {
+						self.finalized_header_listeners.push(tx)
+					},
+					Some(RpcDispatcherMessage::Request(method, params, response_sender)) => {
+						pending_requests.push(client_manager.create_request(method, params, response_sender));
+					},
+					None => {
+						tracing::error!(target: LOG_TARGET, "RPC client receiver closed. Stopping RPC Worker.");
+						return;
+					}
+				},
+				should_retry = pending_requests.next(), if !pending_requests.is_empty() => {
+					if let Some(Err(req)) = should_retry {
+						if let Err(err) = self.self_sender.try_send(req) {
+							tracing::error!(
+								target: LOG_TARGET,
+								?err,
+								"Unable to retry requests, queue is unexpectedly full."
+							);
+							return;
+						};
+						should_reconnect = true;
```

Comment:
```suggestion
						should_reconnect = ShouldReconnect::Reconnect(req);
```

Then we keep the first failed request around for the next loop round.